### PR TITLE
Bicycle improvement: Add support for tag "segregated"

### PR DESCRIPTION
**Pull request by ratrun**
Implementation for #656.
Additionally now assume network "lcn" for a bicycle relation in case that the network tag is missing.

### CODE REVIEW

**Comment by karussell** [comment]
Thanks! Looks like a useful improvement!

> Additionally now assume network "lcn" for a bicycle relation in case that the network tag is missing.

Why this?

Would you mind to avoid formatting changes like adding or removing spaces?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 1]
```diff
@@ -385,6 +387,12 @@ int getSpeed( OSMWay way )
         int speed = PUSHING_SECTION_SPEED;
         String highwayTag = way.getTag("highway");
         Integer highwaySpeed = highwaySpeeds.get(highwayTag);
+        // Override highway speed for otherwise classified as pushing sections, in case that the tagging is similar to highway=cycleway
+        if (way.hasTag("highway", pushingSectionsHighways) && !highwayTag.equals("steps"))
+        {
+            if ((way.hasTag("foot", "no")) || (way.hasTag("segregated", "yes") || (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official"))))
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: I would remove foot=no here as this does not imply (like the others) that bike is designated

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 2]
```diff
@@ -413,32 +421,38 @@ int getSpeed( OSMWay way )
...
             {
-                if (!way.hasTag("service"))
-                    speed = highwaySpeed;
-                else
-                    speed = highwaySpeeds.get("living_street");
+                    if (!way.hasTag("service"))
+                        speed = highwaySpeed;
+                    else
+                        speed = highwaySpeeds.get("living_street");
+                }
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: E.g. here is just formatting changes which should be avoided. Maybe we should implement an automated code styler. For NetBeans and IntelliJ we have some configs already.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 3]
```diff
@@ -413,32 +421,38 @@ int getSpeed( OSMWay way )
...
-                    speed = PUSHING_SECTION_SPEED * 2;
-            }
-        }
+                {
+                   if (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official"))
+                   {
+                       if (way.hasTag("foot", intendedValues) && !way.hasTag("segregated", "yes"))
+                           speed = PUSHING_SECTION_SPEED * 2;
+                   }
+                   else
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: please remove brackes only if a "one-liner-if" statement

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 4]
```diff
@@ -550,7 +564,8 @@ void collect( OSMWay way, double wayTypeSpeed, TreeMap<Double, Integer> weightTo
     {
         String service = way.getTag("service");
         String highway = way.getTag("highway");
-        if (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official"))
+        if ( ((highway!=null) && !highway.equals("steps")) && 
+              (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official")) )
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: if it is designated for bike then we do not need this steps check IMO? Also why is this related to segregated?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 5]
```diff
@@ -571,19 +589,34 @@ void collect( OSMWay way, double wayTypeSpeed, TreeMap<Double, Integer> weightTo
...
+                    weightToPrioMap.put(50d, AVOID_AT_ALL_COSTS.getValue());
+            }
 
         if (pushingSectionsHighways.contains(highway)
                 || way.hasTag("bicycle", "use_sidepath")
                 || "parking_aisle".equals(service))
         {
-            if (way.hasTag("bicycle", "yes"))
-                weightToPrioMap.put(100d, UNCHANGED.getValue());
+            if (way.hasTag("bicycle", intendedValues) && !highway.equals("steps"))
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: same here, why exclude 'steps'?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 2 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Thanks. In my diff the whitespace diff was turned off, so I didn't notice these changes.

**Comment by ratrun** [comment]
>>Additionally now assume network "lcn" for a bicycle relation in case that the network tag is missing.

>Why this?

Occasionally one can find relations with the following tagging: `type=route, route= bicycle` without a network tag. Currently we do not increase the preference for these relations, which can be improved by assuming `network=lcn`.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 1 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Thanks for detecting. -> Fixed.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 4 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: >it is designated for bike then we do not need this steps check IMO? If it is tagged as steps & designated this seems more like a tagging bug

Exactly. And therefore I now do not push the priority. 

> Also why is this related to segregated?

It is true, the `step` change is unrelated to `segregated`. But I would like to keep this.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 5 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: See above. I would like to keep this.

**Comment by ratrun** [comment]
Sorry for the space formating problem. Now I included the changes anew in a new commit and overwrote the first attempt.

**Comment by karussell** [comment]
> Sorry for the space formating problem. Now I included the changes anew in a new commit and overwrote the first attempt.

No real problem, we should automate this formatting somehow before pushing...

>> Also why is this related to segregated?
> It is true, the step change is unrelated to segregated. But I would like to keep this.

Would you mind separating this 'steps' thing into another PR? This way we later have it easier to find out the cause in case of troublesome routing etc. Also having too many things in one PR delay merging. E.g. I would merge this now without the steps thing, as I'm not too convinced it improves the situation and should be discussed separately.

**Comment by ratrun** [comment]
Ok. Now the step handling is removed, I did this via the git web interface. In case that you need the changes squashed into one commit, this will take until next week.

**Comment by ratrun** [comment]
Now the changes are squashed into one commit. Could you please check once again?

**Comment by ratrun** [comment]
What happened with this PR that there is no progress?

**Comment by karussell** [comment]
Sorry for this. The problem I did not immediately merge this is due to several side changes not directly related to the PR. Will check again tomorrow and make this statement more precise.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 6]
```diff
@@ -571,19 +589,34 @@ void collect( OSMWay way, double wayTypeSpeed, TreeMap<Double, Integer> weightTo
...
+                    weightToPrioMap.put(50d, AVOID_AT_ALL_COSTS.getValue());
+            }
 
         if (pushingSectionsHighways.contains(highway)
                 || way.hasTag("bicycle", "use_sidepath")
                 || "parking_aisle".equals(service))
         {
-            if (way.hasTag("bicycle", "yes"))
-                weightToPrioMap.put(100d, UNCHANGED.getValue());
+            if (way.hasTag("bicycle", intendedValues))
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Can this be simplified? Why is this complex if necessary for just a few cases like pushing sections etc?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 7]
```diff
@@ -559,7 +574,10 @@ void collect( OSMWay way, double wayTypeSpeed, TreeMap<Double, Integer> weightTo
         }
         
         if ("cycleway".equals(highway))
-            weightToPrioMap.put(100d, VERY_NICE.getValue());
+            if (way.hasTag("foot", intendedValues) && !way.hasTag("segregated", "yes"))
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Can you put brakets around the parent if to make the flow clear (although logic should be the same)?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 8]
```diff
@@ -550,7 +565,7 @@ void collect( OSMWay way, double wayTypeSpeed, TreeMap<Double, Integer> weightTo
     {
         String service = way.getTag("service");
         String highway = way.getTag("highway");
-        if (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official"))
+        if ((highway!=null) && (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official")))
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Can the highway be empty?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java [hunk 9]
```diff
@@ -434,12 +444,17 @@ int getSpeed( OSMWay way )
             } else 
             {
                 if (way.hasTag("bicycle", "designated") || way.hasTag("bicycle", "official"))
-                    speed = getHighwaySpeed("cycleway");
-                else                
-                    speed = PUSHING_SECTION_SPEED * 2;
-            }
-        }
-
+                {
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: In these cases here we (also) have just pushing sections ... is the complex if handling really necessary or can you simplify this?

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 6 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Sorry, I cannot follow you From my perspective the code complexity remains the same compared to ther situation before the change. 
If the comment is meant as a general complaint about the complexity of the pushing section handling, which I believe it is, then I do not now how it could be simplified without negative impacts on the route quality.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 7 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Thanks. I'll change that.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 8 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: I checked. You are right. As this does not make sense OSMI provides a check for it, so apperently this is accepted the OSM API -> I added that check.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 9 above)
Comment (ratrun) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: I'll think about this here. It is too long ago. I attempted to simplify it now but suddenly some totally unrelated tests fail, althouth the change did pass Travis when I created the PR.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 6 above)
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: > From my perspective the code complexity remains the same compared to ther situation before the change. 

I mean the next 15 lines

> If the comment is meant as a general complaint about the complexity of the pushing section handling

It is really important for me that I do never "complain"! I was just asking. Of course complexity is sometimes necessary and here I wanted to find out if the difference for the route quality is really worth the effort of the nested ifs. E.g. there are two statements `weightToPrioMap.put(100d, VERY_NICE.getValue());` which could be merged to avoid nesting a bit.

File: core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java (hunk 9 above)
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/BikeCommonFlagEncoder.java]: Maybe it was just a hickup on travis side and you try locally again?

**Comment by ratrun** [comment]
I have re-implemented all changes based on yesterdays master and all the feedback given for the original PR above. The local build works for me, I have no idea what "appveyor", whatever it is good for, is complaining about now. Could you please check and give me some hints?

**Comment by ratrun** [comment]
Now all changes are integrated and the introduced complexity is at least better explainsed. And finally also the tests pass. Please check.

File: core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java [hunk 10]
```diff
@@ -70,33 +70,68 @@ public void testGetSpeed() {
         way.clearTags();
         way.setTag("highway", "footway");
         way.setTag("bicycle", "yes");
+        assertEquals(PUSHING_SECTION_SPEED, encoder.getSpeed(way));
+        //assertPriority(PREFER.getValue(), way);
```
Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java]: Is this uncommented intended?

**Comment by karussell** [comment]
Thanks again, looks good! Will just do some minor changes and see again the Harsdorf test (locally to me ;)) and then merge it.

**Comment by karussell** [comment]
Merged now, where I put a TODO to the integration test RoutingAlgorithmWithOSMIT.testHarsdorf
